test/kubernetes/deployment.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networking "k8s.io/api/networking/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+// EchoService name of the deployment for the echo app
+const EchoService = "echo"
+
+// EchoContainer container image name
+//const EchoContainer = "gcr.io/k8s-staging-ingressconformance/echoserver@sha256:e4e132da40d303f4b50e183b977ed1eb5c6db6eb0ddbdaa19565d14d49e05940"
+const EchoContainer = "aledbf/echoserver"
+
+const deploymentTemplate = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: %v
+spec:
+  replicas: 1
+  strategy:
+    type: RollingUpdate
+  selector:
+    matchLabels:
+      app: %v
+  template:
+    metadata:
+      labels:
+        app: %v
+    spec:
+      containers:
+      - name: ingress-conformance-echo
+        image: %v
+        env:
+        - name: POD_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.name
+        - name: NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        - name: INGRESS_NAME
+          value: %v
+        - name: SERVICE_NAME
+          value: %v
+        ports:
+        - name: %v
+          containerPort: 3000
+        readinessProbe:
+          httpGet:
+            path: /health
+            port: 3000
+`
+const serviceTemplate = `
+apiVersion: v1
+kind: Service
+metadata:
+  name: %v
+spec:
+  selector:
+    app: %v
+  ports:
+    - port: %v
+      targetPort: 3000
```

Comment:
Not a big deal, but I usually enjoy having named `targetPort` instead of numbers. It makes migrations and changes easier... although probably irrelevant in this case.

Reply:
I wrote this reading the .feature files. Port is mandatory and only some Ingresses have a name.